torch/csrc/distributed/c10d/NCCLUtils.hpp

```diff
@@ -150,26 +160,38 @@ class NCCLComm {
   }
 
   ncclComm_t getNcclComm() {
-    std::unique_lock<std::mutex> lock(mutex_);
     if (aborted_) {
-      TORCH_CHECK(false,
-          "NCCL communicator was aborted on rank " + std::to_string(rank_) +
-          ".");
+      std::string abortMsg = "NCCL communicator was aborted on rank " + std::to_string(rank_) + ".";
+      if (commFailureReason_ != c10::nullopt) {
+        abortMsg += " . Original reason for failure was: " + *commFailureReason_;
+      }
+      TORCH_CHECK(false, abortMsg);
     }
     return ncclComm_;
   }
 
-  void ncclCommAbort() {
+  c10::optional<std::string> getNcclCommFailureReason() const {
+    std::unique_lock<std::mutex> lock(mutex_);
+    return commFailureReason_;
+  }
+
+  void ncclCommAbort(
+      c10::optional<std::string> commFailureReason = c10::nullopt) {
     std::unique_lock<std::mutex> lock(mutex_);
 #ifdef ENABLE_NCCL_ERROR_CHECKING
     if (aborted_) {
       // Should not abort twice.
       return;
     }
 
-    C10D_NCCL_CHECK(::ncclCommAbort(ncclComm_));
+    C10D_NCCL_CHECK(::ncclCommAbort(ncclComm_), commFailureReason_);
```

Comment:
This is using the member variable `commFailureReason_` instead of the one passed in?

Reply:
Oh right, it is not set until L193

Reply:
Although, I don't think it really matters. Passing in `commFailureReason_` here is not strictly necessary, as we are doing a `ncclCommAbort` i.e. trying to abort the communicator. Wrapping it in C10d_nccl_check will only cause `commFailureReason_` to be raised if something went wrong in aborting the communicator, which is a case I don't think I've seen.

Reply:
Actually, I am not completely sure how we should think about passing in `commFailureReason` to c10d_nccl_check. The main motivation is to ensure all appropriate cases can consider this error when running through the `getNcclErrorDetailStr` error formatter, although a lot of these calls to C10D_NCCL_CHECK are NCCL APIs that don't really have to do with collectives (i.e. ncclCommInitRank, ncclCommAbort, doesn't really have to do anything with the collective itself). 

The calls to C10D_NCCL_CHECK that are indeed collectives (such as the ones in ProcessGroupNCCL::collective) should probably have commFailureReason passed in to ensure we raise the appropriate error if PG work encountered an error though. For the rest of the calls, the benefit is that we'd raise a PyTorch/PG NCCL error where possible instead of a generic NCCL error, which might be better for debugability since we can see what went wrong in our application with regard to timeouts etc.

Reply:
I think we should pass in the `commFailureReason` to all calls which deal with the communicator irrespective of whether or not it is a collective. Because once the communicator is aborted, other nccl calls which might not necessarily be collectives (ex: ncclCommAbort) might probably fail. As a result, I do think we need to pass in the local `commFailureReason` here and not the member variable.